client/asset/dcr/dcr.go

```diff
@@ -5119,28 +5126,142 @@ func (dcr *ExchangeWallet) SetVSP(url string) error {
 
 // PurchaseTickets purchases n number of tickets. Part of the asset.TicketBuyer
 // interface.
-func (dcr *ExchangeWallet) PurchaseTickets(n int, feeSuggestion uint64) ([]*asset.Ticket, error) {
+func (dcr *ExchangeWallet) PurchaseTickets(n int, feeSuggestion uint64) error {
 	if n < 1 {
-		return nil, nil
+		return nil
 	}
 	if !dcr.connected.Load() {
-		return nil, errors.New("not connected, login first")
+		return errors.New("not connected, login first")
 	}
 	// I think we need to set this, otherwise we probably end up with default
 	// of DefaultRelayFeePerKb = 1e4 => 10 atoms/byte.
 	feePerKB := dcrutil.Amount(dcr.feeRateWithFallback(feeSuggestion) * 1000)
 	if err := dcr.wallet.SetTxFee(dcr.ctx, feePerKB); err != nil {
-		return nil, fmt.Errorf("error setting wallet tx fee: %w", err)
+		return fmt.Errorf("error setting wallet tx fee: %w", err)
+	}
+
+	remain := dcr.ticketBuyer.remaining.Add(int32(n))
+	dcr.emit.Data(ticketDataRoute, &TicketPurchaseUpdate{Remaining: uint32(remain)})
+	go dcr.runTicketBuyer()
+
+	return nil
+}
+
+const ticketDataRoute = "ticketPurchaseUpdate"
+
+// TicketPurchaseUpdate is an update from the asynchronous ticket purchasing
+// loop.
+type TicketPurchaseUpdate struct {
+	Err       string          `json:"err,omitempty"`
+	Remaining uint32          `json:"remaining"`
+	Tickets   []*asset.Ticket `json:"tickets"`
+}
+
+// runTicketBuyer attempts to buy requested tickets. Because of a dcrwallet bug,
+// its possible that (Wallet).PurchaseTickets will purchase fewer tickets than
+// requested, without error. To work around this bug, we add requested tickets
+// to ExchangeWallet.ticketBuyer.remaining, and re-run runTicketBuyer every
+// block.
+func (dcr *ExchangeWallet) runTicketBuyer() {
+	tb := &dcr.ticketBuyer
+	if !tb.running.CompareAndSwap(false, true) {
+		// already running
+		return
+	}
+	defer tb.running.Store(false)
+	var ok bool
+	defer func() {
+		if !ok {
+			tb.remaining.Store(0)
+		}
+	}()
+
+	if tb.unconfirmedTickets == nil {
+		tb.unconfirmedTickets = make(map[chainhash.Hash]struct{})
+	}
+
+	remain := tb.remaining.Load()
+	if remain < 1 {
+		return
+	}
+
+	var unconf int
```

Comment:
Don't really need this variable. Can get unconfirmed count later via `len(tb.unconfirmedTickets)`.